### PR TITLE
[Perf] Qwen3-TTS: cut per-step host work in the AR decode loop

### PR DESCRIPTION
**Pull request by JiaxinD**
## Motivation

Qwen3-TTS 1.7B single-instance serving plateaus far below what the card sustains: the same H200 runs a same-card DP2+MPS pool ~1.5x faster (#1418). py-spy at c32 shows the AR scheduler thread burning 72% of one core in per-step Python work (repetition-penalty set rebuilds, suppress-token tensor rebuilds, sampling-buffer restaging, per-row output clones), which is why the ceiling is size independent (0.6B and 1.7B peak identical).

## Changes

All host-side, no numerics change:

* `model_runner/base.py`: repetition-penalty unique-token tracking is incremental per request (was a full `unique(output_ids)` rebuild per step, quadratic over generation; shrunk `output_ids` resets state). Codec suppress-token device tensors cached by list identity (was per-step Python loop + tensor build).
* `models/qwen3_tts/sglang_model.py`: `prepare_decode_buffers` reuses staged sampling state via a request-id fingerprint when batch composition is unchanged (pattern from qwen3_omni talker #1043); predictor-graph capture invalidates the fingerprint.
* `models/qwen3_tts/model_runner.py`: one batched output snapshot clone per step instead of two clones per row (pattern from #1167); feedback staging stacks straight into the graph buffer (pattern from #1439); decode row ids cached.
* `models/qwen3_tts/stages.py` + `scheduling/threaded_simple_scheduler.py`: preprocessing runs on `ThreadedSimpleScheduler` (pattern from higgs/fish/moss_tts_local); the threaded scheduler gains the `abort_callback` contract the qwen3 prepared-request registry cleanup requires (invoked on abort and on late finish after abort).
* `models/qwen3_tts/streaming_vocoder.py`: the per-decode `torch.cat` rebuilt the full stream history although the decode window only moves forward; dead frames behind `window_start` are pruned and slices translate by the offset, making each decode O(window) instead of O(stream). Windows verified byte-identical to full-history slicing across a simulated long stream.

## Evidence (one H200, EN-1088, Seed-TTS harness, main `25ae3d95` base)

Serving defaults for the measured rows: `--max-running-requests 64 --cuda-graph-max-bs 64 --talker-torch-compile-max-bs 64 --isolate-stage vocoder` (defaults retune tracked in #1418; measured numbers below marked measured, everything else inferred).

Single-pass grids (measured). Attribution: the serving flags alone (no code change, unlocked by #1413) already move the peak; this branch's own delta on top of those flags is the row-over-row difference below, concentrated at c32.

| qps | c8 | c16 | c32 | c64 | c96 | peak | delta vs baseline | C16 EN WER |
|---|---|---|---|---|---|---|---|---|
| baseline defaults | 6.08 | 8.40 | 8.74 | 7.44 | 8.29 | 8.74 | — | 0.99% |
| + admission 64 + vocoder isolate, main code | 6.26 | 8.30 | 8.85 | 11.96 | 12.05 | 12.05 | +37.9% | 0.99% |
| + this branch (same flags, pre-hardening head) | 6.44 | 8.83 | 10.98 | 10.28 | 12.62 | 12.62 | +44.4% | pending |
| DP2+MPS same silicon (#1400 recipe, MF=0.40) | 4.57 | 7.05 | 9.43 | 10.29 | 10.02 | 10.29 | +17.7% | n/a |

This branch over the flags-only row: c32 +24.1% (8.85 -> 10.98), peak +4.7% (12.05 -> 12.62); the flags rows carry the rest of the cumulative gain. Defaults re-tune to make the flags row the out-of-the-box behavior is tracked in #1418 section 2.

H100 calibration of the final head, one card back to back, quiet box (load 3.8-8), all cells 1088/1088:

| qps | c8 | c16 | c32 | c64 | c96 | peak | TTFA p95 @c64 |
|---|---|---|---|---|---|---|---|
| baseline defaults | 7.23 | 10.33 | 10.64 | 10.71 | 10.40 | 10.71 | 4.76 s |
| this branch + flags | 8.16 | 11.31 | 13.81 | 15.80 | 16.15 | **16.15 (+50.8%)** | **0.54 s** |
| DP2+MPS re-measured (#1400 recipe) | 7.14 | 11.13 | 14.81 | 13.81 | 15.66 | 15.66 | 2.80 s |

The optimized single instance now **exceeds** the same-card DP2+MPS pool peak (16.15 vs 15.66, +3.2%) while holding a 5x better TTFA p95 at c64, so the #1418 exit criterion is met outright rather than approached: colocation is no longer the recommended deployment for this variant at peak throughput, and it was never competitive on first-audio latency. DP2 still leads at c32 alone (14.81 vs 13.81). C16 EN WER on the optimized head is 1.12% against the 1.07% published baseline, inside the 0.2pt gate.

TTFA p95 stays flat at c8/c16 in every row (0.20-0.34s); with admission 64 the c64 cliff is gone (7.77s -> 0.69-0.97s).

Same-card three-arm ABAB, 5 rounds, H200, peak qps medians over c32-96: baseline 8.74, this branch + flags 12.40 (+42%, four clean rounds; one round lost to the pre-existing intermittent c64 device assert tracked in #1418 section 4), DP2+MPS 13.76. TTFA p95 at c64 medians: 0.95s vs 2.91s vs 5.93s. On H200 the branch lands at 90% of the pool peak with a 3x TTFA advantage; on H100 (table below) it lands at 95.2%.

Rebased head (adds the vocoder window pruning, current main base), single pass on H200, all cells 1088/1088: 6.34 / 8.90 / 10.95 / 11.50 / 10.96 qps at c8-c96, within the ABAB round spread above.

## Follow-up work in this branch

Three further changes landed after the first review round, each measured separately:

* **Streaming vocoder window pruning.** The per-decode `torch.cat` rebuilt the whole stream history although the window only moves forward; dead chunks are dropped and slices translate by the pruned offset. Windows verified byte-identical to full-history slicing.
* **Codec-id screen.** An id outside the codebook reached the embedding lookup and became a device-side assert that poisoned the CUDA context and killed every in-flight stream in the vocoder process (`validate_chunk` skips device tensors to avoid a per-chunk sync). The decoder input is now clamped into range and the per-row verdict is read back after the decode's existing `stream.synchronize()`, so no decode gains a synchronization, and only the offending rows fail while the rest of the batch still decodes.
* **Device-resident shaping masks + pinned id staging.** Repetition penalty and codec suppression move from per-step host index building to persistent device masks updated by a single scatter of the last sampled token (rebuilt whenever a history does not grow by exactly one token, so retract and restart stay correct); sampled ids are staged into pinned host memory so the output processor's `.tolist()` waits on an event instead of issuing a blocking pageable copy.

Paired A/B on one H200, three alternating rounds, masks on vs off (same head otherwise):

| c16 qps | r1 | r2 | r3 | mean |
|---|---|---|---|---|
| masks off | 8.61 | 8.51 | 8.63 | 8.58 |
| masks on | 9.44 | 9.26 | 9.49 | **9.40 (+9.5%)** |

At c32 the same comparison is neutral (11.11 vs 11.09), so the win is concentrated where the AR scheduler thread is the binding constraint. The pinned staging was A/B'd the same way and is **not** credited with a gain: three rounds gave -0.9%, +5.6%, +4.7% at c16, so any effect is below this setup's ~5% noise floor. It is kept because it is strictly less host work on the same path qwen3_omni already uses, not because it measured faster. The tradeoff is measurable and consistent: c16 TTFA p95 rises from 0.36s to 0.42s (+17%) as batches run deeper. Full-grid single pass with everything on: 6.95 / 8.97 / 11.42 / 13.04 / 13.53 qps at c8-c96, zero failures, which is 98% of the same-card DP2+MPS pool peak on that box. Overload behaviour improves too: c64 +22% qps over the masks-off head (13.04 vs 10.65) with TTFA p95 1.08s, and c96 TTFA p95 is better rather than worse (2.87s vs 3.74s) with lower p99 latency throughout. C16 EN WER on that head is 1.04% against 0.99% for the same-box baseline arms, inside the 0.2pt quality gate.

## Tests

* `tests/unit_test/qwen3_omni/test_logit_shaping.py`: incremental repetition penalty equivalence across growth and retract-shrink steps; suppress-token grouped/cached path equivalence vs scalar reference (shared and per-row lists, out-of-range ids, repeated calls through the cache).
* `tests/unit_test/qwen3_tts/test_pipeline.py`: abort-race cleanup test adapted to the threaded executor, same contract (late finish after abort still cleans the prepared registry); stream-prune window test asserting pruned decode windows stay byte-identical to full-history slicing while chunks actually get dropped.
* `tests/unit_test/qwen3_tts/test_mask_logit_shaping.py`: mask shaping equals the scalar reference across steady steps, request-id reuse, batch shrink, history shrink after a retract, and an emptied history.
* `tests/unit_test/qwen3_tts/test_pipeline.py`: a decode batch with one out-of-range row fails exactly that row, keeps the decoder from ever seeing the bad id, and still decodes the remaining rows.
* Full `qwen3_tts` + `scheduling` + `model_runner` + `qwen3_omni` unit suites green (191 in the TTS-facing subset after the follow-up work; 463 across the wider set earlier in the branch).

## Cross-model disclosure

`model_runner/base.py` and `threaded_simple_scheduler.py` are shared surfaces: the penalty/suppress helpers also serve the thinker path (qwen3_omni), and the threaded scheduler is used by higgs/fish/moss_tts_local preprocessing. The suppress/penalty changes are output-equivalent by unit test; the scheduler change is additive (new optional parameter, default None keeps prior behavior).

## Acknowledgements

Thanks to @vuuihc for the original bounded ThreadedSimpleScheduler abort-marker cleanup in #611. This PR incorporated that earlier work while adding the abort_callback contract; I missed the cross-reference when #1462 was merged.

### CODE REVIEW

File: sglang_omni/model_runner/base.py [hunk 1]
```diff
@@ -946,13 +977,30 @@ def _apply_codec_suppress_tokens(self, logits_output: Any, requests: list) -> No
...
+            else:
+                toks = [int(t) for t in suppress_tokens if 0 <= int(t) < vocab]
+                toks_t = (
+                    torch.tensor(toks, dtype=torch.long, device=device)
+                    if toks
+                    else None
+                )
+                if len(cache) >= 64:
+                    cache.clear()
+                cache[key] = (suppress_tokens, toks_t)
```
Comment (0xjeffro) [inline sglang_omni/model_runner/base.py]: To my understand, no matter how the concurrency, it looks like this cache should only hold one element?
so what is this cap `64` guarding for?

File: sglang_omni/model_runner/base.py [hunk 2]
```diff
@@ -933,8 +958,14 @@ def _apply_codec_suppress_tokens(self, logits_output: Any, requests: list) -> No
...
+        # filtered device tensor by list identity makes every decode step after
+        # a request's first a cache hit instead of a rebuild.
+        cache = getattr(self, "_suppress_tensor_cache", None)
+        if cache is None:
+            cache = {}
+            self._suppress_tensor_cache = cache
+        row_groups: dict[int, tuple[torch.Tensor, list[int]]] = {}
         for row_idx, sched_req in enumerate(requests):
             data = sched_req.data
             suppress_tokens = data.suppress_tokens
```
Comment (0xjeffro) [inline sglang_omni/model_runner/base.py:972]: nit: `data.suppress_token` is passed in from request builder, but logically this is a model config, so this should identical for every request. maybe we can just built once when model runner init, rather than cache them?

File: sglang_omni/model_runner/base.py (hunk 1 above)
Comment (JiaxinD) [reply sglang_omni/model_runner/base.py]: Right, the cap was covering for identity keying: the builder hands out a fresh list per request, so it churned one entry per request. Now keyed by content, so it holds one entry and the cap is gone.

File: sglang_omni/model_runner/base.py (hunk 2 above)
Comment (JiaxinD) [reply sglang_omni/model_runner/base.py:972]: Agreed it is model config; base cannot assume that for every model, so the content key is derived once per request and the tensor is shared fleet-wide. Same effect as building at init, without the assumption.

**Review by 0xjeffro — APPROVED** [review]

**Comment by JiaxinD** [comment]
Thanks @vuuihc for catching this and for the original work in #611. I have updated the PR description to credit that earlier patch; the missing cross-reference was my oversight.
